automerge/src/op_observer.rs

```diff
@@ -1,50 +1,105 @@
 use crate::exid::ExId;
+use crate::Parents;
 use crate::Prop;
 use crate::Value;
 
 /// An observer of operations applied to the document.
-pub trait OpObserver {
+pub trait OpObserver: Default + Clone {
     /// A new value has been inserted into the given object.
     ///
     /// - `objid`: the object that has been inserted into.
     /// - `index`: the index the new value has been inserted at.
     /// - `tagged_value`: the value that has been inserted and the id of the operation that did the
     /// insert.
-    fn insert(&mut self, objid: ExId, index: usize, tagged_value: (Value<'_>, ExId));
+    fn insert(
+        &mut self,
+        parents: Parents<'_>,
+        objid: ExId,
+        index: usize,
+        tagged_value: (Value<'_>, ExId),
+    );
 
     /// A new value has been put into the given object.
     ///
     /// - `objid`: the object that has been put into.
-    /// - `key`: the key that the value as been put at.
+    /// - `prop`: the prop that the value as been put at.
     /// - `tagged_value`: the value that has been put into the object and the id of the operation
     /// that did the put.
     /// - `conflict`: whether this put conflicts with other operations.
-    fn put(&mut self, objid: ExId, key: Prop, tagged_value: (Value<'_>, ExId), conflict: bool);
+    fn put(
+        &mut self,
+        parents: Parents<'_>,
+        objid: ExId,
+        prop: Prop,
+        tagged_value: (Value<'_>, ExId),
+        conflict: bool,
+    );
 
     /// A counter has been incremented.
     ///
     /// - `objid`: the object that contains the counter.
-    /// - `key`: they key that the chounter is at.
+    /// - `prop`: they prop that the chounter is at.
     /// - `tagged_value`: the amount the counter has been incremented by, and the the id of the
     /// increment operation.
-    fn increment(&mut self, objid: ExId, key: Prop, tagged_value: (i64, ExId));
+    fn increment(
+        &mut self,
+        parents: Parents<'_>,
+        objid: ExId,
+        prop: Prop,
+        tagged_value: (i64, ExId),
+    );
 
     /// A value has beeen deleted.
     ///
     /// - `objid`: the object that has been deleted in.
-    /// - `key`: the key of the value that has been deleted.
-    fn delete(&mut self, objid: ExId, key: Prop);
+    /// - `prop`: the prop of the value that has been deleted.
+    fn delete(&mut self, parents: Parents<'_>, objid: ExId, prop: Prop);
+
+    /// Merge data with an other observer
+    ///
+    /// - `other`: Another Op Observer of the same type
+    fn merge(&mut self, other: &Self);
```

Comment:
It's not clear from the docs when an observer would be merged or branched. I think it's worth adding a few sentences to explain when this would happen for both this method and `branch`.